pkg/parser/parser.peg

```diff
@@ -313,6 +313,65 @@ InlineAttributes <- "[" attrs:(GenericAttribute)* "]" {
     return types.NewAttributes(attrs.([]interface{}))								
 } 
 
+// Quoted Text attributes (attributes in front of marked up text).
+//
+// There are two syntaxes, a regular form, and another special short-hand form.
+// They behave quite differently from each other, and quite differently from the format
+// used for attribute lists elsewhere.  Both are *extremely* limited.
+//
+// The regular form consists *solely* of a role.  It is the entire text between
+// the open and closing brackets.  No transforms or validation are done.  If this
+// contains illegal characters, asciidoctor passes it straight-through, leading to
+// invalid context.
+//
+// Note that just to make matters more fun, the handling of these lists is somewhat
+// different for constrained vs unconstrained elements.  With constrained elements,
+// the presence of a comma or named attributes is not tolerated, and the list is
+// passed to the output.  With unconstrained markup, any named attributes are simply
+// ignored (everything after the comma is just discarded).  For the sake of sanity
+// we elect to use the more generous form and just discard everything after and including
+// the first comma.  (Perhaps someday more sanity will prevail, and will let us use
+// real named values, including id=, in this form.)  It appears that asciidoctor will
+// trim trailing whitespace, but not leading.  Another win for consistency! (Or not.)
+//
+// The second form is a short-hand that consists of either .role or #id elements,
+// concatenated together with no whitespace permitted between them, or between the
+// enclosing brackets at all.  Anything except a closing bracket (]), dot (.), or hash (#)
+// may be present in these.  Multiple roles may be present, and if so they will be
+// joined with a space and placed in the CSS class.  (Asciidoctor does not escape these,
+// making it possible to create invalid or insecure content.)
+//
+// Corollary 1: There can only be role or ID attributes.
+// Corollary 2: The ID cannot contain a hash or a dot.
+// Corollary 3: There can be only one role if it contains either a hash or a dot.
+// Corollary 4: Neither the ID nor the role may contain a closing bracket.
+//
+// Conclusion: Limit oneself to [0-9A-Za-z_-] in roles and IDs for reliable results.
```

Comment:
I like the idea of limiting ourselves to `[0-9A-Za-z_-]+` in roles and IDs, but I can't see it in the grammar rules below :(

Reply:
It's not in the rules.   Rather this is a recommendation for document authors -- the results will vary across implementations when using other characters.  I've tried to be as generous as we can reasonably allow.

See Postel's Law.